library/core/src/str/traits.rs

```diff
@@ -198,15 +199,31 @@ unsafe impl const SliceIndex<str> for ops::Range<usize> {
         let slice = slice as *const [u8];
         // SAFETY: the caller guarantees that `self` is in bounds of `slice`
         // which satisfies all the conditions for `add`.
-        let ptr = unsafe { slice.as_ptr().add(self.start) };
+        let ptr = unsafe {
+            let this = ops::Range { ..self };
+            assert_unsafe_precondition!(
+                "str::get_unchecked requires that the range is within the string slice",
+                (this: ops::Range<usize>, slice: *const [u8]) =>
+                this.end >= this.start && this.end <= slice.len()
```

Comment:
`str::get_unchecked` also has the precondition that the indexes must be on char boundaries, just like the checked version. we should verify that here too since that's perhaps trickier to get right than staying in bounds.

Reply:
Previous discussion: https://github.com/rust-lang/rust/pull/105117#discussion_r1058479207

There's not really a way to do so without reading from behind the pointer, which allegedly has aliasing implications.

Reply:
Ah well, then maybe a comment why that isn't done here?

Reply:
That's fair, I'll add a comment.